File: src/64.最小路径和.js
/*
 * @lc app=leetcode.cn id=64 lang=javascript
 *
 * [64] 最小路径和
 */

// @lc code=start
/**
 * @param {number[][]} grid
 * @return {number}
 */
var minPathSum = function(grid) {
  for (let i = 1 ; i < grid[0].length; i++) {
    grid[0][i] = grid[0][i] + grid[0][i-1]
  }
  for (let i = 1 ; i < grid.length; i++) {
    grid[i][0] = grid[i][0] + grid[i-1][0]
  }

  for (let i = 1 ; i < grid.length; i++) {
    for (let j = 1 ; j < grid[0].length; j++) {
      grid[i][j] = Math.min(grid[i-1][j], grid[i][j-1]) + grid[i][j]
    }
  }
  console.log(grid)
  return grid[grid.length - 1][grid[0].length - 1]
};
// @lc code=end
/**
 * 1 3 1
 * 1 5 1
 * 4 2 1
 * 
 * 1 4 5
 * 2 7 6
 * 6 8 7
 */

File: src/64.最小路径和.rs
/*
 * @lc app=leetcode.cn id=64 lang=rust
 *
 * [64] 最小路径和
 */

// @lc code=start
impl Solution {
    pub fn min_path_sum(grid: Vec<Vec<i32>>) -> i32 {
        let mut dp = vec![];
        for i in 0..grid.len() {
            dp.push(vec![]);

            for j in 0..grid[i].len() {
                if i == 0 && j == 0 {
                    dp[i][j] = grid[0][0] as usize;
                } else if i == 0 {
                    dp[i][j] = dp[i][j-1] + (grid[i][j] as usize);
                } else if j == 0 {
                    dp[i][j] = dp[i-1][j] + (grid[i][j] as usize);
                } else {
                    dp[i][j] = std::cmp::min(dp[i-1][j], dp[i][j-1]) + (grid[i][j] as usize);
                }
            }
        }

        dp[dp.len() - 1][dp[0].len() - 1] as i32
    }
}
// @lc code=end

